### PR TITLE
Add support for standard library std::backtrace::Backtrace

### PR DESCRIPTION
**Pull request by ccfb3ee765a58cae**
See #172. This gives a proper `.backtrace()` to snafu errors, using nightly's `#![feature(backtrace)]`.

When enabled, this re-exports `std::backtrace::Backtrace` as `snafu::Backtrace` and changes `snafu::ErrorCompat::backtrace` such that it returns a std library backtrace instead of one from the `backtrace` crate.

May be a bit messy, as I haven't contributed to this project before. There are some considerations that need to be decided (see git commit comments) before this is merged, so I'm leaving it as a draft.

### CODE REVIEW

File: snafu-derive/src/lib.rs [hunk 1]
```diff
@@ -1296,7 +1297,11 @@ impl<'a> quote::ToTokens for ContextSelector<'a> {
         let backtrace_field = match *backtrace_field {
             Some(ref field) => {
                 let name = &field.name;
-                quote! { #name: std::default::Default::default(), }
+                if cfg!(feature = "std-backtraces") {
+                    quote! { #name: snafu::Backtrace::capture(), }
```
Comment (ccfb3ee765a58cae) [inline snafu-derive/src/lib.rs]: Should we use force_capture or capture? Just using capture means the backtrace will only be created if RUST_BACKTRACE=1 at runtime. See #84

File: src/lib.rs [hunk 2]
```diff
@@ -441,7 +445,7 @@ impl<T> OptionExt<T> for Option<T> {
 /// ```
 pub trait ErrorCompat {
```
Comment (ccfb3ee765a58cae) [inline src/lib.rs]: Should we even bother exposing an `ErrorCompat` when this feature is enabled?

Keeping it enabled seems like it would allow the smoothest ride to future rust.

File: compatibility-tests/std-backtrace/tests/backtrace.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
+#![feature(backtrace)]
```
Comment (ccfb3ee765a58cae) [inline compatibility-tests/std-backtrace/tests/backtrace.rs]: These tests are just copied and pasted from `tests/`. Tests specific to snafu with `std::backtrace::Backtrace` support may be useful.

File: snafu-derive/src/lib.rs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #![allow(unknown_lints, bare_trait_objects)]
 #![recursion_limit = "128"] // https://github.com/rust-lang/rust/issues/62059
+#![cfg_attr(feature = "std-backtraces", feature(backtrace))]
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: Why is this needed? I would have thought it isn't, as we are only generating code that would use this feature, not using it ourselves.

File: src/lib.rs (hunk 2 above)
Comment (shepmaster) [reply src/lib.rs]: I think so, yes. For as long as we want to support backtraces on stable Rust before the standard library stabilizes, we'll need to have something. If we were to remove it, then the compatibility shim wouldn't be providing much value to those users.

File: snafu-derive/src/lib.rs (hunk 1 above)
Comment (shepmaster) [reply snafu-derive/src/lib.rs]: Yeah, this is something I'd struggled a bit with when thinking through the problem. I think that using `capture` is the right place to start; we might eventually add something like `#[snafu(backtrace(force))]` to allow opting to the other function.

Comment (yaahc) [reply snafu-derive/src/lib.rs]: I just left a comment on this in another repo, but you could have it interpret `Backtrace` as meaning `force_capture` and `Option<Backtrace>` as a regular `capture`

**Comment by shepmaster** [comment]
I hope to get to this soon! Please see [my comment](https://github.com/shepmaster/snafu/pull/157#issuecomment-539511054) about my plans.

**Comment by shepmaster** [comment]
Here's [my proposed combination of the two branches](https://github.com/shepmaster/snafu/compare/backtrace-revamp), plus some reorganizing, refactoring, and small feature additions.

TL;DR:

- `snafu::Backtrace` always exists (thanks to @lucab)
- The feature flag `backtraces` makes the `Backtrace` type live and implemented by our own type. 
- The feature flag `backtraces-impl-backtrace-crate`  makes the `Backtrace` type live and a re-export of `backtrace::Backtrace`.
- The feature flag `unstable-backtraces-impl-std`  makes the `Backtrace` type live and a re-export of `std::backtrace::Backtrace` (thanks to @ccfb3ee765a58cae).
- There's a new trait for generating backtraces to abstract across these three implementations and the trait allows `Option<Backtrace>`, if you want to prevent capturing the backtrace unless an environment variable is set (thanks to @yaahc).

How do people feel about this direction?

**Comment by shepmaster** [comment]
Thanks!
